File: Leet 1477/solution.cs
Console.WriteLine("1477. Find Two Non-overlapping Sub-arrays Each With Target Sum");
Console.WriteLine();

int result = Solution.MinSumOfLengths([3,2,2,4,3], 3);
Console.WriteLine($"MinSumOfLengths([3,2,2,4,3], 3) should equal 2 ? {result}");

class Solution 
{
    public static int MinSumOfLengths(int[] arr, int target) 
    {
        int[] dp = new int[arr.Length + 1];
        for(int i = 0; i <= arr.Length; i++)
        {
            dp[i] = arr.Length + 1;
        }

        int result = arr.Length + 1;
        int total = 0;

        int left = 0;
        for(int right = 0; right < arr.Length; right++)
        {
            total += arr[right];
            while(total > target)
            {
                total -= arr[left++];
            }

            dp[right + 1] = dp[right];

            if (total == target)
            {
                int currentWindow = right - left + 1;

                dp[right + 1] = Math.Min(dp[right + 1], currentWindow);
                result = Math.Min(result, dp[left] + currentWindow);
            }
        }

        return result == arr.Length + 1 ? -1 : result;
    }
}

File: Leet 1477/solution.c
#include <stdlib.h>

#define min(a, b) (((a) < (b)) ? (a) : (b))

int minSumOfLengths(int* arr, int arrSize, int target) 
{
    int maxWindowSize = arrSize + 1;

    int* dp = calloc(sizeof(int), arrSize + 1);
    for (int i = 0; i <= arrSize; i++)
    {
        dp[i] = maxWindowSize;
    }

    int result = maxWindowSize;
    int total = 0;
    
    int left = 0;
    
    for(int right = 0; right < arrSize; right++)
    {
        total += arr[right];

        while (total > target)
        {
            total -= arr[left++];
        }

        dp[right + 1] = dp[right];

        if (total == target)
        {
            int currentWindow = right - left + 1;
            dp[right + 1] = min(dp[right + 1], currentWindow);

            result = min(result, dp[left] + currentWindow);
        }
    }

    free(dp);
    
    return result != maxWindowSize ? result : -1;
}

